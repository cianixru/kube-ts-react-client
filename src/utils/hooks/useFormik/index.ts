// @credits: https://egghead.io/lessons/react-build-your-own-formik-using-react-hooks-with-jared-palmer
import { useReducer } from 'react';
import { SET_FIELD_TOUCHED, SET_FIELD_VALUE } from './actions';

export interface FormikAction {
  readonly type: string;
  readonly payload?: any;
}

export const formReducer = <T>(state: T, action: FormikAction) => {
  switch (action.type) {
    default:
      return state;
  }
};

export interface Options<T> {
  readonly initialValues: T;
  readonly onSubmit: (values: T) => void;
}

export type Touched<T> = { readonly [P in keyof T]: boolean };

export type Errors<T> = { readonly [P in keyof T]: string | undefined | null };

export interface FormikHook<Values> {
  readonly handleBlur: (event: React.ChangeEvent<HTMLInputElement>) => void;
  readonly handleChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
  readonly handleSubmit: (event: Event) => void;
  readonly values: Values;
  readonly touched: Touched<Values>;
  readonly errors: Errors<Values>;
}

function useFormik<Values>({
  initialValues,
  onSubmit,
}: Options<Values>): FormikHook<Values> {
  const [state, dispatch] = useReducer(formReducer, {
    errors: {},
    touched: {},
    values: initialValues,
  });

  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    event.persist();

    const { name, value } = event.target;

    dispatch({ type: SET_FIELD_VALUE, payload: { name, value } });
  };

  const handleBlur = (event: React.ChangeEvent<HTMLInputElement>) => {
    event.persist();

    const { name } = event.target;

    dispatch({ type: SET_FIELD_TOUCHED, payload: { name } });
  };

  const handleSubmit = (event: Event) => {
    event.preventDefault();

    onSubmit(state.values);
  };

  return {
    errors: state.errors,
    handleBlur,
    handleChange,
    handleSubmit,
    touched: state.touched,
    values: state.values,
  };
}

export default useFormik;
