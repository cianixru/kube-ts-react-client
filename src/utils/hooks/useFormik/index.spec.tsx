// TODO: move that (jest-dom/extend-expect) to setup  file
// tslint:disable-next-line:no-import-side-effect
import '@testing-library/jest-dom/extend-expect';
// TODO: move that (@testing-library/react/cleanup-after-each) to setup file
import { render } from '@testing-library/react';
import React from 'react';
import useFormik from './index';

interface Values {
  readonly email: string;
  readonly password: string;
}

describe('@useFormik', () => {
  const defaultOptions = {
    initialValues: {
      email: '',
      password: '',
    },
    onSubmit: jest.fn(),
  };

  const renderHook = () => {
    let result: any;

    const HookWrapper = () => {
      result = useFormik<Values>(defaultOptions);

      return null;
    };

    render(<HookWrapper />);

    return result;
  };

  it('returns initial state', () => {
    const result = renderHook();

    expect(result).toMatchSnapshot();
  });
});
