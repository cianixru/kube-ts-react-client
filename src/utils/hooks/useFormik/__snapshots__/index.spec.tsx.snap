// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@useFormik returns initial state 1`] = `
Object {
  "errors": Object {},
  "handleBlur": [Function],
  "handleChange": [Function],
  "handleSubmit": [Function],
  "touched": Object {},
  "values": Object {
    "email": "",
    "password": "",
  },
}
`;
